src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1124,17 +974,329 @@
    "ALTER TABLE fact_paths DROP COLUMN value_type_id"
    "ALTER TABLE fact_values DROP COLUMN path_id"))
 
-(defn fill-in-null-producer-timestamp
-  "The producer_timestamp column can sometimes contain null values,
-  especially when updating from older versions. New use cases require
-  a value in that position, so we'll use the value from the
-  'timestamp' column if it's not there. "
+(defn version-2yz-to-300-migration
+  ;; This migration includes:
+  ;;   Insertion of the factsets hash column
+  ;;   Using a report_id (surrogate key) instead of the reports.hash (natural key)
+  ;;   Insert a noop column into reports
+  ;;   Drop latest_report table, merge with certnames
+  ;;   Change name to certname where inconsistent
+  ;;   Insert reports metrics and logs
+  ;;   Ensuring producer_timestamp is NOT NULL
+  ;;   Changing the types of hashes and uuids in postgres to bytea and uuid respectively
```

Comment:
@wkalt To some degree I would agree, things like Russ's timestamp patch or adding the noop column or renaming name to certname we didn't really get any benefit of collapsing here, it just obscured that fact those things were actually happening. The only reason I collapsed those was because I really wanted to collapse this migration with the report_id migration and since I was reordering things anyway it seemed more reasonable to collapse things. But when we have to drop our table/transform/rebuild our tables I would say there could be real performance concerns if we duplicate that procedure across migrations in the same release, and any benefit of keeping things discrete should be weighed against the complexity of duplicating migration code in a single release. Like @kbarber said though I think the major concerns of things like this is lack of testing around it.